Quantum-Control-Applications/ESR/cpmg.py

```diff
@@ -0,0 +1,213 @@
+from qm import SimulationConfig
+from qm.qua import *
+from qm import LoopbackInterface
+from qm.QuantumMachinesManager import QuantumMachinesManager
+from configuration import *
+import matplotlib.pyplot as plt
+from qm.simulate.credentials import create_credentials
+
+###################
+# The QUA program #
+###################
+
+tau_min = 2000 // 4  # tau is half the time between pi pulses
+tau_max = 5e4 // 4  # tau is half the time between pi pulses
+tau_array = np.round(np.exp(np.linspace(np.log(tau_min), np.log(tau_max), 30)))  # outputs results in float
+
+tau_array_int = tau_array.astype(int).tolist()
+
+num_taus = len(tau_array)
+
+latest_echo = 50e6 // 4  # this is the maximum amount of time I will run CPMG
+
+num_echo_array = (np.floor(latest_echo / (2 * tau_array))).astype(int).tolist()
+
+total_num_echoes = np.sum(num_echo_array)  # total number of echoes played in the whole sequence
+
+n_avg = 100
+
+with program() as cpmg:
+
+    n = declare(int)
+    n_st = declare_stream()
+
+    I = declare(fixed)
+    Q = declare(fixed)
+
+    I_st = declare_stream()
+    Q_st = declare_stream()
+
+    tau_st = declare_stream()
+    i_echo_st = declare_stream()
+
+    tau_array_q = declare(int, value=tau_array_int)
+    tau = declare(int)
+    i_tau = declare(int)
+
+    num_echo_array_q = declare(int, value=num_echo_array)
+    i_echo = declare(int)
+
+    echo = declare_stream(adc_trace=True)
+
+    pulse1_len = declare(int, value=240 // 4)  # 250 cycles = 1 us
+    pulse2_len = declare(int, value=480 // 4)
+
+    with for_(n, 0, n < n_avg, n + 1):
+
+        with for_(i_tau, 0, i_tau < num_taus, i_tau + 1):
+
+            # initialization
+            # wait(cooldown_time)
+
+            reset_phase("ensemble")
+            reset_phase("resonator")
+            reset_frame("ensemble")
+
+            assign(tau, tau_array_q[i_tau])
+
+            wait(11, "switch_1", "switch_2")
```

Comment:
Where does this value come from?

Reply:
so we want to cover 5 ms of measurement and we chose the values to do the experiment on so that they are equal space in log scale spannign these 5 ms that is why